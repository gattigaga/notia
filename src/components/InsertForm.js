import React from "react";
import PropTypes from "prop-types";
import styled from "styled-components";
import TextareaAutosize from "react-autosize-textarea";
import moment from "moment";

const Container = styled.div`
  width: 480px;
  background: white;
  box-shadow: 2px 2px 2px rgba(0, 0, 0, 0.1);
  margin: 32px auto;
  padding: 16px;
  padding-bottom: 8px;
`;

const InputTitle = styled(TextareaAutosize)`
  width: 100%;
  font-weight: bold;
  font-family: Roboto;
  font-size: 16px;
  color: black;
  border: 0px;
  outline: 0px;
  resize: none;
  margin-bottom: 12px;

  &::placeholder {
    color: #888;
    user-select: none;
  }
`;

const Toolbar = styled.div`
  width: 100%;
  padding-top: 16px;
  display: flex;
  justify-content: flex-end;
`;

const Button = styled.button`
  padding: 8px 14px;
  border: 0px;
  border-radius: 2px;
  background: white;
  font-family: Roboto;
  font-size: 12px;
  font-weight: 500;
  text-transform: uppercase;
  color: #333;
  cursor: pointer;
  outline: 0px;

  &:hover {
    background: #eee;
  }
`;

const Date = styled.p`
  color: #aaa;
  font-size: 12px;
  font-family: Roboto;
  font-style: italic;
  margin: 0px;
  margin-top: 16px;
  text-align: right;
`;

Button.displayName = "Button";
InputTitle.displayName = "InputTitle";

const InsertForm = ({
  className,
  title,
  mode,
  timestamp,
  children,
  onClickDone,
  onChangeTitle,
  onFocusTitle
}) => {
  const date = moment(timestamp).format("LLLL");

  return (
    <Container className={className}>
      <InputTitle
        rows={1}
        placeholder="Title"
        value={title}
        onChange={onChangeTitle}
        onFocus={onFocusTitle}
      />
      {children}
      {mode === "edit" && <Date>Edited at {date}</Date>}
      <Toolbar>
        <Button onClick={onClickDone}>DONE</Button>
      </Toolbar>
    </Container>
  );
};

InsertForm.propTypes = {
  className: PropTypes.string,
  onClickDone: PropTypes.func,
  onChangeTitle: PropTypes.func,
  onFocusTitle: PropTypes.func,
  children: PropTypes.node,
  title: PropTypes.string,
  mode: PropTypes.oneOf(["create", "edit"]),
  timestamp: PropTypes.number
};

InsertForm.defaultProps = {
  mode: "create",
  timestamp: 0
};

export default InsertForm;
