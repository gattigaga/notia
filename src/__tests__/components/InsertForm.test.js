import React from "react";
import { shallow } from "enzyme";
import toJSON from "enzyme-to-json";

import InsertForm from "components/InsertForm";

let wrapper, onClickDone, onChangeTitle, onFocusTitle;

beforeEach(() => {
  onClickDone = jest.fn();
  onChangeTitle = jest.fn();
  onFocusTitle = jest.fn();

  wrapper = shallow(
    <InsertForm
      title="Learn React"
      onClickDone={onClickDone}
      onChangeTitle={onChangeTitle}
      onFocusTitle={onFocusTitle}
    />
  );
});

afterEach(() => {
  onClickDone.mockReset();
  onChangeTitle.mockReset();
  onFocusTitle.mockReset();
});

describe("InsertBox", () => {
  it("renders correctly", () => {
    expect(toJSON(wrapper)).toMatchSnapshot();
  });

  it("renders with additional class", () => {
    wrapper.setProps({ className: "my-class" });
    expect(toJSON(wrapper)).toMatchSnapshot();
  });

  it("calls 'onClickDone'", () => {
    expect(onClickDone.mock.calls.length).toEqual(0);
    wrapper.find("Button").simulate("click");
    expect(onClickDone.mock.calls.length).toEqual(1);
  });

  it("calls 'onFocusTitle'", () => {
    expect(onFocusTitle.mock.calls.length).toEqual(0);
    wrapper.find("InputTitle").simulate("focus");
    expect(onFocusTitle.mock.calls.length).toEqual(1);
  });

  it("calls 'onChangeTitle'", () => {
    const event = {
      target: {
        value: "Learn React"
      }
    };

    expect(onChangeTitle.mock.calls.length).toEqual(0);
    wrapper.find("InputTitle").simulate("change", event);
    expect(onChangeTitle.mock.calls.length).toEqual(1);
    expect(onChangeTitle).toBeCalledWith(event);
  });
});
