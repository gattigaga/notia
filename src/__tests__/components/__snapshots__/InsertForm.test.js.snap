// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InsertBox renders correctly 1`] = `
<styled.div>
  <InputTitle
    onChange={[Function]}
    onFocus={[Function]}
    placeholder="Title"
    rows={1}
    value="Learn React"
  />
  <styled.div>
    <Button
      onClick={[Function]}
    >
      DONE
    </Button>
  </styled.div>
</styled.div>
`;

exports[`InsertBox renders with additional class 1`] = `
<styled.div
  className="my-class"
>
  <InputTitle
    onChange={[Function]}
    onFocus={[Function]}
    placeholder="Title"
    rows={1}
    value="Learn React"
  />
  <styled.div>
    <Button
      onClick={[Function]}
    >
      DONE
    </Button>
  </styled.div>
</styled.div>
`;
